### ISSUE TITLE
batch versioning

### ISSUE THREAD

**Issue by denislepage** [MotusDev/Motus-TO-DO#437]
I took a few steps towards versioning.

I have created a new table called sg_versions and added a record with version = 1. I have included a field called “available” which is meant to indicate which version are still being supported, and a ts field indicating when the version started. We will probably want a few more things in there. The current version should always be the max(version) in that table.

I have added a new field called version to the sg_batches table. Once the scripts are operational, I’ll want to make that a required field.

I have modified the import scripts in sandbox so the highest version number is read from sg_versions at the start of an import loop, and that value is added to the sg_batches table. The changes have been posted on sandbox tonight, so we’ll see if that worked with the next import.

Next steps:

In database, for views that need filtering (e.g. the daily summaries), I suggest that they rely on a query rather than having a version number being hard-coded (select max(version) as version from sg_version). 

For R views, which will not be filtered themselves, the API code will need to obtain the default version value during initialization, and support overriding the default with an earlier version. We may want to add an API call to return the list of supported versions available for users. 

We’ll need to carefully review our table indexes for queries doing any sort of filtering based on version. We should keep a list of the views and processes being modified as we go, so we can evaluate and optimize indexes. We likely will have to modify only one or 2 indexes on sg_batches in the end.

No need to rush on these, but I thought I would get us started. Baby steps.

**Comment by denislepage** [comment]
I further added the version field as an output in the vwR_batches... views. It may be one of those fields that we remove from the JSON response output, but not just yet. The main possible reason to keep it is if we plan to allow users to mix different versions of data in the same SQLite file.  I'll talk to Steffi this week, she may offer some insight. There would be ways to prevent this, but there are a few considerations. 

The other views and products should not provide the version in their output, rather simply use it to filter them. 

I tried to create a new function that gives us the current SG version (max(version) from sg_versions). 

This is very fast: select dbo.SGVersion()

I would have expected these 2 following queries to run in about the same amount of time, but not so:

This is very fast:
select top 10000 * from vwR_Batches_for_tag_project  a
where batchID > 0 and  motusProjectID = 1 and version = 1
order by a.batchID

But this is slow (~5 seconds):
select top 10000 * from vwR_Batches_for_tag_project  a
where batchID > 0 and  motusProjectID = 1 and version = dbo.SGVersion()
order by a.batchID

I expect the function needs to be called for every row (10.,000 times) in the latter case (non-deterministic function), which is so very not efficient.

In procs, this approach is also fast:

DECLARE @version int = dbo.SGVersion()
select top 10000 * from vwR_Batches_for_tag_project  a
where batchID > 0 and  motusProjectID = 1 and version = @version
order by a.batchID
GO

Finally, this approach is also fast:
select top 10000 * from vwR_Batches_for_tag_project  a
where batchID > 0 and  motusProjectID = 1 and version = (select MAX(version) from sg_versions where available = 1) order by a.batchID

In the R API, we can load the version number once at initialization time and use that as the default unless explicitly specified by the user.

In conclusion, it sounds like we won't be able to use the convenience of a UDF (user defined function) to return the version number as I hoped, at least not within views.

**Comment by denislepage** [comment]
Found this: https://www.brentozar.com/archive/2014/10/sql-server-functions-dragging-query/

Worth a read regarding the use (or NOT) and the impact of scalar functions in queries. It confirms my suspicions about why the above query was slower.

**Comment by denislepage** [comment]
I added some code in sandbox for data versioning (it was only partly in place). The highest version number is now correctly being added to the sg_batches table during the SG data import. 

Once we have confirmed that it works ok (so far so good), we'll need to : 1) bring the changes in SGDataImport from sandbox to beta, deploy to beta, and add a new entry in sg_api_fields. We can't do this step in the database until the Java code has been deployed.

```
INSERT INTO sg_api_fields (table_name, remote_field, local_field)
VALUES ('sg_batches','version','version')

```

I'll have to confirm at some point, but I think that the version number will be read at the start of every loop and used within a job. Creating a new version number should therefore be detected by the next SGDataImport job, with no other operations needed.

**Comment by Hussell** [comment]
SGDataImport has begun failing on beta with the following error message:
java.lang.IllegalArgumentException: Invalid field name: version
   at org.bsc.motus.process.SGDataImport.writeTable(SGDataImport.java:1102)
   at org.bsc.motus.process.SGDataImport.readBatchesForAll(SGDataImport.java:767)
   at org.bsc.motus.process.SGDataImport.loop(SGDataImport.java:471)
   at org.bsc.motus.process.SGDataImport.run(SGDataImport.java:196)
   at org.bsc.motus.process.Job.run(Job.java:379)

I deployed to beta this morning, but I don't see any obvious code changes in the update that might be related to this.

**Comment by denislepage** [comment]
This was the certainly the result of adding an entry in sg_api_fields for the sg_taghits_deploy_id table. I wasn’t sure if it were going to create this problem, but I guess I have my answer.

The field also exists in the live database, but there has to be something I missed when I changed the database without updating the beta code.

Solution: merge SGDataImport from master into beta (I think the changes are working there) or perhaps temporarily remove the version value from sg_api_fields for sg_taghits_deploy.

I would go with the first option, I think. It should not matter that the changes aren’t immediately deployed to live, since the import is only happening on beta.

We should fix this today, as this is blocking new data from coming in.

**Comment by Hussell** [comment]
Ok, I can do this and redeploy to beta. It looks like there are no unrelated changes to SGData in master.

**Comment by Hussell** [comment]
Now live is failing with the error:
java.lang.ClassCastException: java.lang.Integer cannot be cast to org.json.simple.JSONArray
   at org.bsc.motus.process.SGDataImport.writeTable(SGDataImport.java:1112)
   at org.bsc.motus.process.SGDataImport.readBatchesForAll(SGDataImport.java:778)
   at org.bsc.motus.process.SGDataImport.loop(SGDataImport.java:480)
   at org.bsc.motus.process.SGDataImport.run(SGDataImport.java:205)
   at org.bsc.motus.process.Job.run(Job.java:379)

Checking back to the sandbox queue, I see that's been failing with this error too.

**Comment by denislepage** [comment]
Yes, there was a cast error that I fixed yesterday, but since the field was not listed in the db table, it was not triggered.

Presumably, this is fixed now in sandbox and beta?

I can deploy to live if the changes are working in beta. Is this blocking any from working? I don’t think that class is being used in live.

**Comment by Hussell** [comment]
No, the above error has been happening in sandbox since I deployed this morning, and is happening in beta now that it matches sandbox. Have you pushed your fix?

**Comment by denislepage** [comment]
Nope. ☺

Done now. I had done an ant build to test it, but didn’t push my commits.
